### ISSUE TITLE
[BUG] Upload same VM image again when Longhorn still deleting previous causes status stuck in Uploading

### ISSUE THREAD

**Issue by albinsun** [harvester/harvester#4293]
## Description
Upload same VM image when Longhorn still deleting the previous one causes status stuck in Uploading.

## Environment
 - Harvester ISO version: `v1.2.0-rc4`
 - Underlying Infrastructure: QEMU/KVM, single node (8C/16G/250G)

## Reproduce Steps
Not easy reproduce from UI but in automation.

Automation steps (via API):
1. Upload VM image with unique name
2. Wait until GET VM Image API reply OK
3. Delete VM image
4. Wait until GET VM Image API reply 404
5. Upload same VM image again with same name
6. Upload API reply 504 timeout, image status stuck in Uploading
   <details>

   ![image](https://github.com/harvester/harvester/assets/2773781/6c751f75-81f0-4919-ab2f-c30a72b10506)
   ![image](https://github.com/harvester/harvester/assets/2773781/774ba3d4-715e-485f-a25d-7e86b966158f)

   </details>

[record.webm](https://github.com/harvester/harvester/assets/2773781/92e269b1-4b9b-4376-a2bb-d8d929b56cbe)

### Reference
https://github.com/albinsun/tests/blob/upload_delete_twice/harvester_e2e_tests/integration/test_images.py#L195

### Trigger
```
$ pwd
/home/john/harvester/tests

$ tox -e py36 --result-json=test_result.json -- -l harvester_e2e_tests  --json-report --html=test_report/test_result.html -m 'not delete_host and not upgrade' -k 'delete_image_recreate'
```

## Expected behavior
For Harvester API user, 2nd upload should be success since the 1st image is deleted. (But in fact LH is still deleting)

### Note
1. If we wait 1st deletion in LH done (sleep ~10s), TC will pass.
2. The InProgress 1st deletion seems affects 2nd upload in the same time.
3. Related issue: harvester#4027

## Support bundle
[supportbundle_2beb3a27-a21a-49de-a654-97c56a5514fa_2023-07-24T01-41-42Z.zip](https://github.com/harvester/harvester/files/12141242/supportbundle_2beb3a27-a21a-49de-a654-97c56a5514fa_2023-07-24T01-41-42Z.zip)

**Comment by WebberHuang1118** [comment]
TL;DR, in this e2e testing case, after deleting an image, it creates an image with the same name and has a POST to upload immediately. However, the `old bi` is still under deletion, and the `new bi` has not been created yet. This leads to the error `failed backing image is deleted, requesting manager to clean up the tmp file of backing image data source"` during [waiting `bids` ready](https://github.com/harvester/harvester/blob/1de4c2f8fd6f9214b313d39fab5e568c535fcd42/pkg/api/image/formatter.go#L165-L167).

**Reproduce Steps**
- Build a Harvester with master-head (`1de4c2f8fd6f9214b313d39fab5e568c535fcd42`)
- Running following e2e testing
  - `$ tox -e py36 --result-json=test_result.json -- -l harvester_e2e_tests  --json-report --html=test_report/test_result.html -m 'not delete_host and not upgrade' -k 'delete_image_recreate or test_create_image_url'`

**Investigation**
From the [custom Harvester](https://github.com/WebberHuang1118/harvester/tree/issue-4293) branch with detailed log:
```
time="2024-07-16T08:00:27Z" level=info msg="vmi 15h59m05s410374-07-16 create bi default-15h59m05s410374-07-16"
....
....
time="2024-07-16T08:00:46Z" level=info msg="vmi 15h59m05s410374-07-16 trying to upload image"
time="2024-07-16T08:00:46Z" level=info msg="vmi 15h59m05s410374-07-16 ds default-15h59m05s410374-07-16 is ready"
time="2024-07-16T08:00:47Z" level=info msg="vmi 15h59m05s410374-07-16 uploadResp ready"
time="2024-07-16T08:00:47Z" level=info msg="vmi 15h59m05s410374-07-16 upload ready"
time="2024-07-16T08:00:47Z" level=info msg="vmi 15h59m05s410374-07-16 delete bi default-15h59m05s410374-07-16"
time="2024-07-16T08:00:52Z" level=info msg="vmi 15h59m05s410374-07-16 trying to upload image"
time="2024-07-16T08:00:52Z" level=info msg="vmi 15h59m05s410374-07-16 waiting bids ready failed backing image is deleted, requesting manager to clean up the tmp file of backing image data source"
time="2024-07-16T08:00:57Z" level=info msg="vmi 15h59m05s410374-07-16 create bi default-15h59m05s410374-07-16"
```

old image `15h59m05s410374-07-16` is deleted at `2024-07-16T08:00:47Z`, then a new image `15h59m05s410374-07-16` is created and receives POST upload at `2024-07-16T08:00:52`. However, because the old bi is still under the deletion process, the POST had the error `backing image is deleted, requesting manager to clean up the tmp file of backing image data source`. The new bi creation was delayed to `2024-07-16T08:00:57`.

**Comment by WebberHuang1118** [comment]
This issue could be solved by PR https://github.com/harvester/harvester/pull/5114, since we will name `bi` by `vmi`'s UID rather than the `<namespace-name>` combination, so even if the image is recreated with the same name, Harvester controller wouldn't try to access the `old bi` resource

Testing result:
![4293_testing_result](https://github.com/user-attachments/assets/84c9858a-65be-4b0a-a0e4-fe63f6af4cae)

**Comment by w13915984028** [comment]
A question here:

Looks the `bi` has no `OwnerReferences` to `image` is the root cause, instead of changing the name, could we utilize the 
 `OwnerReferences` to pair those objects?

**Comment by WebberHuang1118** [comment]
Hi @w13915984028 

I think the `bi` has no `OwnerReferences` for some specific reason from the beginning design. Will update if there is any finding, thanks.

**Comment by w13915984028** [comment]
Guess one reason is that the backingimage objects are put in `longhorn-system` namespace while the `virtualmachineimage` are put in specific namespace. It breaks the ownerreferences in general. 

Not sure why the namespaces are separated.

Besides the strong binding of names, did you consider to annotate the virtualmachineimage info into backingimage? Just like the ownerreferences, but instead via an annotation. It will keep consistency.

```
$kubectl  get backingimage -A
NAMESPACE         NAME                  UUID       SOURCETYPE   SIZE        VIRTUALSIZE   AGE
longhorn-system   default-image-7qznl   61e99f0f   upload       666763264   2361393152    16h
longhorn-system   default-image-hd6p6   92658268   upload       16300544    117440512     24h
```

```
$kubectl get virtualmachineimage -A
NAMESPACE   NAME          DISPLAY-NAME                             SIZE        VIRTUALSIZE   AGE
default     image-7qznl   ubuntu-22.04-server-cloudimg-amd64.img   666763264   2361393152    16h
default     image-hd6p6   cirros-0.5.2-x86_64-disk.img             16300544    117440512     24h
```

**Comment by WebberHuang1118** [comment]
> Guess one reason is that the backingimage objects are put in longhorn-system namespace while the virtualmachineimage are put in specific namespace. It breaks the ownerreferences in general.

Yes @w13915984028, I just want to update the same information, thanks.

> Besides the strong binding of names, did you consider to annotate the virtualmachineimage info into backingimage? Just like the ownerreferences, but instead via an annotation. It will keep consistency.

We will introduce PR https://github.com/harvester/harvester/pull/5114 to connect `image` and `bi` through UID, which is designated to solve the `bi` naming conflict (like this issue).

**Comment by WebberHuang1118** [comment]
Just double-check with the master-head branch, works as expected. Moving status to ready-for-testing.

**Comment by brandboat** [comment]
Test OK on [v1.4.0-dev-20240918](https://github.com/harvester/harvester/releases/tag/v1.4.0-dev-20240918), close as fixed.

## Environment
* [`v1.4.0-dev-20240918`](https://github.com/harvester/harvester/releases/tag/v1.4.0-dev-20240918)
  * Profile: 2 nodes QEMU/KVM (8C/16G/250G per node)
  * ui-source: `Auto`

## Test Steps
- Loop `tox -e python3.11 --result-json=test_result.json -- -l harvester_e2e_tests --json-report --html=test_report/test_result.html -m 'not delete_host and not upgrade' -k 'delete_image_recreate or create_image_url'` 20 times and all good.
